lib/server/services/user/createOrUpdateUser.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright CERN and copyright holders of ALICE O2. This software is
+ * distributed under the terms of the GNU General Public License v3 (GPL
+ * Version 3), copied verbatim in the file "COPYING".
+ *
+ * See http://alice-o2.web.cern.ch/license for full licensing information.
+ *
+ * In applying this license CERN does not waive the privileges and immunities
+ * granted to it by virtue of its status as an Intergovernmental Organization
+ * or submit itself to any jusdiction.
+ */
+
+const { getUser } = require('./getUser.js');
+const { BadParameterError } = require('../../errors/BadParameterError.js');
+const UserRepository = require('../../../database/repositories/UserRepository.js');
+
+/**
+ * Creates or updates a user in the database and returns the user object
+ *
+ * @param {Object} newUser the user to create
+ * @param {number} [newUser.personid] the CERN user's id, used as externalUserId
+ * @param {string|null} [newUser.name] username for in bookkeeping db
+ * @param {string|null} [newUser.access] access level
+ * @param {string|null} [newUser.username] username from cern
+ * @return {Promise<SequelizeUser>} resolve once the creation is done providing the user that has been (or will be) created
+ */
+exports.createOrUpdateUser = async (newUser) => {
+    const { personid, name, access, username } = newUser;
+
+    // Check null OR undefined because !personid returns true with personid == 0
+    if (personid === null || personid === undefined) {
+        throw new BadParameterError('Cannot create without personid');
+    }
+
+    // GetUser expects an identifier with two options, so surround it with {}
+    let user = await getUser({ externalUserId: personid });
+    if (user) {
+        if (user.name !== name) {
+            user = await UserRepository.update(user, { id: user.id, externalId: personid, name });
+        }
+    } else {
+        user = await UserRepository.insert({ externalId: personid, name });
+    }
+
+    // Set user.access for server startup if present (WebUiServer)
+    if (access) {
+        user.access = access;
+    }
+    // Set user.username for server startup if present (WebUiServer)
+    if (username) {
+        user.username = username;
+    }
```

Comment:
If these info are not persisted in the database, for me they should not go further than the high level service.

Reply:
I am not sure what you mean by this, could you explain what you expect me to change here a bit differently maybe?

Reply:
I think I have found a solution to this, see WebUiServer.js for my implemented solution. The only part this is used is in the call from WebUiServer.js, but those variables are also present in request.session, so I can just move this to WebUiServer.js because no other code is using or going to use that.